go/tasks/pluginmachinery/flytek8s/pod_helper.go

```diff
@@ -104,67 +107,230 @@ func UpdatePod(taskExecutionMetadata pluginsCore.TaskExecutionMetadata,
 	ApplyInterruptibleNodeAffinity(taskExecutionMetadata.IsInterruptible(), podSpec)
 }
 
-// ToK8sPodSpec constructs a pod spec from the given TaskTemplate
-func ToK8sPodSpec(ctx context.Context, tCtx pluginsCore.TaskExecutionContext) (*v1.PodSpec, error) {
-	task, err := tCtx.TaskReader().Read(ctx)
+// BuildRawPod constructs a PodSpec and ObjectMeta based on the definition passed by the TaskExecutionContext. This
+// definition does not include any configuration injected by Flyte.
+func BuildRawPod(ctx context.Context, tCtx pluginsCore.TaskExecutionContext) (*v1.PodSpec, *metav1.ObjectMeta, string, error) {
+	taskTemplate, err := tCtx.TaskReader().Read(ctx)
 	if err != nil {
 		logger.Warnf(ctx, "failed to read task information when trying to construct Pod, err: %s", err.Error())
-		return nil, err
+		return nil, nil, "", err
+	}
+
+	var podSpec *v1.PodSpec
+	objectMeta := metav1.ObjectMeta{
+		Annotations: make(map[string]string),
+		Labels:      make(map[string]string),
 	}
-	if task.GetContainer() == nil {
-		logger.Errorf(ctx, "Default Pod creation logic works for default container in the task template only.")
-		return nil, fmt.Errorf("container not specified in task template")
+	primaryContainerName := ""
+
+	if taskTemplate.GetContainer() != nil {
+		// handles tasks defined by a single container
+		c, err := ToK8sContainer(ctx, tCtx)
+		if err != nil {
+			return nil, nil, "", err
+		}
+
+		primaryContainerName = c.Name
+		podSpec = &v1.PodSpec{
+			Containers: []v1.Container{
+				*c,
+			},
+		}
+	} else if taskTemplate.GetK8SPod() != nil {
+		// handles pod tasks that marshal the pod spec to the k8s_pod task target.
+		if taskTemplate.GetK8SPod().PodSpec == nil {
+			return nil, nil, "", pluginserrors.Errorf(pluginserrors.BadTaskSpecification,
+				"Pod tasks with task type version > 1 should specify their target as a K8sPod with a defined pod spec")
+		}
+
+		err := utils.UnmarshalStructToObj(taskTemplate.GetK8SPod().PodSpec, &podSpec)
+		if err != nil {
+			return nil, nil, "", pluginserrors.Errorf(pluginserrors.BadTaskSpecification,
+				"Unable to unmarshal task k8s pod [%v], Err: [%v]", taskTemplate.GetK8SPod().PodSpec, err.Error())
+		}
+
+		// get primary container name
+		var ok bool
+		if primaryContainerName, ok = taskTemplate.GetConfig()[PrimaryContainerKey]; !ok {
+			return nil, nil, "", pluginserrors.Errorf(pluginserrors.BadTaskSpecification,
+				"invalid TaskSpecification, config missing [%s] key in [%v]", PrimaryContainerKey, taskTemplate.GetConfig())
+		}
+
+		// update annotations and labels
+		if taskTemplate.GetK8SPod().Metadata != nil {
+			objectMeta.Annotations = utils.UnionMaps(objectMeta.Annotations, taskTemplate.GetK8SPod().Metadata.Annotations)
+			objectMeta.Labels = utils.UnionMaps(objectMeta.Labels, taskTemplate.GetK8SPod().Metadata.Labels)
+		}
```

Comment:
I'm probably missing something, we don't seem to assign any values to objectMeta before these lines, why do we need to union?

Reply:
I updated this to merge the `taskTemplate` annotations and labels into the `objectMeta`. Would you prefer just setting them explicitly?